apps/emqx_bridge_gcp_pubsub/src/emqx_bridge_gcp_pubsub_consumer_worker.erl

```diff
@@ -0,0 +1,552 @@
+    topic => emqx_bridge_gcp_pubsub_connector:topic()
+}.
+-type decoded_message() :: map().
+
+-define(HEALTH_CHECK_TIMEOUT, 10_000).
+-define(OPTVAR_SUB_OK(PID), {?MODULE, PID}).
+
+%%-------------------------------------------------------------------------------------------------
+%% API used by `reply_delegator'
+%%-------------------------------------------------------------------------------------------------
+
+-spec pull_async(pid()) -> ok.
+pull_async(WorkerPid) ->
+    gen_server:cast(WorkerPid, pull_async).
+
+-spec process_pull_response(pid(), binary()) -> ok.
+process_pull_response(WorkerPid, RespBody) ->
+    gen_server:cast(WorkerPid, {process_pull_response, RespBody}).
+
+-spec ensure_subscription(pid()) -> ok.
+ensure_subscription(WorkerPid) ->
+    gen_server:cast(WorkerPid, ensure_subscription).
+
+-spec reply_delegator(pid(), binary(), {ok, map()} | {error, timeout | term()}) -> ok.
+reply_delegator(WorkerPid, InstanceId, Result) ->
+    case Result of
+        {error, timeout} ->
+            ?MODULE:pull_async(WorkerPid);
+        {error, Reason} ->
+            ?SLOG(warning, #{
+                msg => "gcp_pubsub_consumer_worker_pull_error",
+                instance_id => InstanceId,
+                reason => Reason
+            }),
+            case Reason of
+                #{status_code := 409} ->
+                    %% the subscription was not found; deleted?!
+                    ?MODULE:ensure_subscription(WorkerPid);
+                _ ->
+                    ?MODULE:pull_async(WorkerPid)
+            end;
+        {ok, #{status_code := 200, body := RespBody}} ->
+            ?MODULE:process_pull_response(WorkerPid, RespBody)
+    end.
+
+%%-------------------------------------------------------------------------------------------------
+%% Debugging API
+%%-------------------------------------------------------------------------------------------------
+
+-spec get_subscription(pid()) -> {ok, map()} | {error, term()}.
+get_subscription(WorkerPid) ->
+    gen_server:call(WorkerPid, get_subscription, 5_000).
+
+%%-------------------------------------------------------------------------------------------------
+%% `ecpool' health check
+%%-------------------------------------------------------------------------------------------------
+
+-spec health_check(pid()) -> boolean().
+health_check(WorkerPid) ->
+    case optvar:read(?OPTVAR_SUB_OK(WorkerPid), ?HEALTH_CHECK_TIMEOUT) of
+        {ok, _} ->
+            true;
+        timeout ->
+            false
+    end.
+
+%%-------------------------------------------------------------------------------------------------
+%% `emqx_resource' API
+%%-------------------------------------------------------------------------------------------------
+
+connect(Opts0) ->
+    Opts = maps:from_list(Opts0),
+    #{
+        ack_retry_interval := AckRetryInterval,
+        bridge_name := BridgeName,
+        connector_state := ConnectorState,
```

Comment:
This is a bit disturbing 😅

Because the state is supposed to be controlled by the connector, but it is "multiplied" several times and passed to workers, where it is no more controlled by the connector.

We seem to use the implicit knowledge that the connector's state is actually some constant settings that are valid as long as the workers are alive. Maybe we should at least reflect this fact in the naming somehow?

Reply:
I've experimented managing the connector as an individual resource and interacting with it via `emqx_resource:simple_{,a}sync_query`, but that introduced some additional difficulties:

- that call still logs "errors" like the 409 status code when we ensure that a subscription exists;
- that also could potentially bump some metrics.

We _could_ make some changes to avoid such side effects as well, but I'm more inclined to simply add a comment in this particular bridge warning to _not_ attempt to change the connector state value in the `on_get_status` callback, for simplicity.  :thinking: :see_no_evil:

Reply:
After our discussion, I've transformed the connector into an opaque client.  I think it's cleaner now.  :beers: